tests/test_simple_pipeline_executor.py

```diff
@@ -59,27 +81,150 @@ def test_from_task_class(self):
+
+    def test_from_pipeline_output_differ(self):
+        """Run pipeline but output definition in registry differs."""
+        executor = self._configure_pipeline(
+            NoDimensionsTestTask.ConfigClass,
+            NoDimensionsTestTask.ConfigClass,
+            storageClass_a="TaskMetadataLike",
+        )
+
+        # Pre-define the "output" storage class to be something that is
+        # like a dict but is not a dict. This will fail unless storage
+        # class conversion is supported in put and get.
+        self.butler.registry.registerDatasetType(
+            lsst.daf.butler.DatasetType(
+                "output",
+                dimensions=self.butler.registry.dimensions.empty,
+                storageClass="TaskMetadataLike",
+            )
+        )
+
+        with self.assertLogs("lsst", level="INFO") as cm:
+            quanta = executor.run(register_dataset_types=True)
+        # a has been told to return a TaskMetadata but will convert to dict.
+        # b returns a dict and that is converted to TaskMetadata on put.
+        self._test_logs(cm.output, "dict", "lsst.pipe.base.TaskMetadata", "dict", "dict")
+
+        self.assertEqual(len(quanta), 2)
+        self.assertEqual(self.butler.get("intermediate"), {"zero": 0, "one": 1})
+        self.assertEqual(self.butler.get("output").to_dict(), {"zero": 0, "one": 1, "two": 2})
+
+    def test_from_pipeline_input_differ(self):
+        """Run pipeline but input definition in registry differs."""
+
+        # This config declares that the pipeline takes a TaskMetadata
+        # as input but registry already thinks it has a StructureDataDict.
+        executor = self._configure_pipeline(NoDimensionsTestConfig2, NoDimensionsTestTask.ConfigClass)
+
+        with self.assertLogs("lsst", level="INFO") as cm:
+            quanta = executor.run(register_dataset_types=True)
+        self._test_logs(cm.output, "lsst.pipe.base.TaskMetadata", "dict", "dict", "dict")
+
         self.assertEqual(len(quanta), 2)
         self.assertEqual(self.butler.get("intermediate"), {"zero": 0, "one": 1})
         self.assertEqual(self.butler.get("output"), {"zero": 0, "one": 1, "two": 2})
 
+    def test_from_pipeline_incompatible(self):
+        """Run pipeline but definitions are not compatible."""
+        executor = self._configure_pipeline(
+            NoDimensionsTestTask.ConfigClass, NoDimensionsTestTask.ConfigClass
+        )
+
+        # Incompatible output dataset type.
+        self.butler.registry.registerDatasetType(
+            lsst.daf.butler.DatasetType(
+                "output",
+                dimensions=self.butler.registry.dimensions.empty,
+                storageClass="StructuredDataList",
+            )
+        )
+
+        with self.assertRaises(ValueError) as cm:
+            executor.run(register_dataset_types=True)
+        self.assertIn("inconsistent with registry definition", str(cm.exception))
```

Comment:
Maybe `assertRaisesRegex` instead of separate test for exception message?

Reply:
Thanks. Had forgotten about that.